zirgen/compiler/stats/OpStats.cpp

```diff
@@ -0,0 +1,379 @@
+// Copyright 2025 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "zirgen/compiler/stats/OpStats.h"
+
+#include "mlir/IR/AsmState.h"
+#include "risc0/fp/fp.h"
+#include "risc0/fp/fpext.h"
+#include "zirgen/Dialect/Zll/IR/IR.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Format.h"
+#include "llvm/Support/ManagedStatic.h"
+#include "llvm/Support/Timer.h"
+#include <random>
+
+#define DEBUG_TYPE "op-stats"
+
+using namespace risc0;
+using namespace zirgen::Zll;
+using namespace mlir;
+namespace cl = llvm::cl;
+
+namespace zirgen {
+
+namespace {
+
+enum class SortOrder { Disabled, Max, Inner, Outer, Flat, Any };
+
+struct OpStatsCLOptions {
+  cl::opt<enum SortOrder> sortOrder{
+      "op-stats",
+      cl::desc("Produce operation statistics with this order"),
+      cl::init(SortOrder::Disabled),
+      cl::values(
+          clEnumValN(SortOrder::Disabled, "disabled", "Do not calculate operation statistics"),
+          clEnumValN(SortOrder::Inner, "inner", "Combined across callees"),
+          clEnumValN(SortOrder::Outer, "outer", "Combined across callers"),
+          clEnumValN(SortOrder::Flat, "flat", "Uncombined"),
+          clEnumValN(SortOrder::Any, "any", "Combined occurrences anywhere in call stack"),
+          clEnumValN(SortOrder::Max, "max", "Maximum of all the metrics"))};
+};
+
+llvm::ManagedStatic<OpStatsCLOptions> clOpts;
+
+// Number of times to run profiling loop
+constexpr size_t kNumIter = 1000 * 1000;
+
+// Prevent a value from being optimized for use in profiling loop.
+template <typename FpT> FpT blackBox(FpT value) {
+#if defined(__clang__)
+  asm volatile("" : "+r,m"(value) : : "memory");
+#else
+  asm volatile("" : "+m,r"(value) : : "memory");
+#endif
+  return value;
+}
```

Comment:
haven't seen this pattern before - appears to be an assembly no-op which implies that it's passing a value through without actually doing anything? Any target-specificity concerns with this? Not used to seeing asm in a platform-independent way, but maybe this counts!?

Reply:
Yeah, I adapted it from `DoNotOptimize` in google's benchmark library.  Updated comment...